web/src/AnalyticsNudge.test.tsx

```diff
@@ -1,67 +1,120 @@
+    mockAnalyticsOptInOnce(false)
 
-it("shows request-in-progress message", () => {
-  const component = shallow(<AnalyticsNudge needsNudge={false} />)
+    render(<AnalyticsNudge needsNudge={true} />)
 
-  component.setState({ requestMade: true })
+    expect(screen.getByLabelText("Tilt analytics options")).toBeInTheDocument()
 
-  expect(component).toMatchSnapshot()
-})
+    userEvent.click(screen.getByRole("button", { name: /nope/i }))
 
-it("shows success message: opt out", () => {
-  const component = shallow(<AnalyticsNudge needsNudge={false} />)
+    await waitFor(() => {
+      userEvent.click(screen.getByRole("button", { name: /dismiss/i }))
+    })
 
-  component.setState({ optIn: false, requestMade: true, responseCode: 200 })
+    expect(screen.queryByLabelText("Tilt analytics options")).toBeNull()
+  })
 
-  expect(component).toMatchSnapshot()
-})
+  it("shows request-in-progress message when a request is in progress", () => {
+    mockAnalyticsOptInOnce(false)
 
-it("shows success message: opt in", () => {
-  const component = shallow(<AnalyticsNudge needsNudge={false} />)
+    render(<AnalyticsNudge needsNudge={true} />)
 
-  component.setState({ optIn: true, requestMade: true, responseCode: 200 })
+    userEvent.click(screen.getByRole("button", { name: /nope/i }))
 
-  expect(component).toMatchSnapshot()
-})
+    expect(screen.getByTestId("opt-loading"))
+  })
+
+  it("shows success message for opt out", async () => {
+    mockAnalyticsOptInOnce(false)
+    render(<AnalyticsNudge needsNudge={true} />)
 
-it("shows failure message with response body", () => {
-  const component = shallow(<AnalyticsNudge needsNudge={false} />)
+    userEvent.click(screen.getByRole("button", { name: /nope/i }))
 
-  component.setState({
-    requestMade: true,
-    responseCode: 418,
-    responseBody: "something is not right! something is quite wrong!",
+    await waitFor(() => {
+      expect(screen.getByTestId("optout-success")).toBeInTheDocument()
+    })
   })
 
-  expect(component).toMatchSnapshot()
-})
+  it("shows success message for opt in", async () => {
+    mockAnalyticsOptInOnce(true)
+    render(<AnalyticsNudge needsNudge={true} />)
 
-it("hidden if dismissed", () => {
-  const component = shallow(<AnalyticsNudge needsNudge={false} />)
+    userEvent.click(screen.getByRole("button", { name: /I'm in/i }))
 
-  component.setState({ requestMade: true, responseCode: 200, dismissed: true })
+    await waitFor(() => {
+      expect(screen.getByTestId("optin-success")).toBeInTheDocument()
+    })
+  })
+
+  it("shows a failure message when request fails", async () => {
+    mockAnalyticsOptInOnce(true, true)
+    render(<AnalyticsNudge needsNudge={true} />)
+
+    userEvent.click(screen.getByRole("button", { name: /I'm in/i }))
+
+    await waitFor(() => {
+      expect(screen.getByRole("alert")).toBeInTheDocument()
+    })
+  })
 
-  expect(component).toMatchSnapshot()
+  // TODO: Fix this test, if possible! There's a timing issue with the
```

Comment:
ya, this is a legit bug!

in OverviewLogPane, we handle this by keeping track of the timer ids, and cancelling them on unmount. we should do something similar here

https://github.com/tilt-dev/tilt/blob/master/web/src/OverviewLogPane.tsx#L301

Reply:
ah, yes! thanks for the pointer. i fixed this and the test is now passing. :tada: